src/macaron/repo_finder/provenance_extractor.py

```diff
@@ -355,3 +358,354 @@ def check_if_repository_purl_and_url_match(url: str, repo_purl: PackageURL) -> b
+        gl_workflow = json_extract(statement["predicate"], ["attestations", "attestation", "ciconfigpath"], str)
+        gl_job_url = json_extract(statement["predicate"], ["attestations", "attestation", "joburl"], str)
+        return gl_workflow, gl_job_url
+
+
+class ProvenancePredicate:
+    """Class providing utility methods for handling provenance predicates.
+
+    This class contains static methods for extracting information from predicates in
+    provenance statements related to various build definitions. It serves as a helper
+    for identifying build types and finding the appropriate build definitions based on the extracted data.
+    """
+
+    @staticmethod
+    def get_build_type(statement: InTotoV1Statement | InTotoV01Statement) -> str | None:
+        """Extract the build type from the provided provenance statement.
+
+        Parameters
+        ----------
+        statement : InTotoV1Statement | InTotoV01Statement
+            The provenance statement from which to extract the build type.
+
+        Returns
+        -------
+        str | None
+            The build type if found; otherwise, None.
+        """
+        if statement["predicate"] is None:
+            return None
+
+        if build_type := json_extract(statement["predicate"], ["buildType"], str):
+            return build_type
+
+        return json_extract(statement["predicate"], ["buildDefinition", "buildType"], str)
+
+    @staticmethod
+    def find_build_def(statement: InTotoV01Statement | InTotoV1Statement) -> ProvenanceBuildDefinition:
+        """Find the appropriate build definition class based on the extracted build type.
+
+        This method checks the provided provenance statement for its build type
+        and returns the corresponding `ProvenanceBuildDefinition` subclass.
+
+        Parameters
+        ----------
+        statement : InTotoV01Statement | InTotoV1Statement
+            The provenance statement containing the build type information.
+
+        Returns
+        -------
+        ProvenanceBuildDefinition
+            An instance of the appropriate build definition class that matches the
+            extracted build type.
+
+        Raises
+        ------
+        ProvenanceError
+            Raised when the build definition cannot be found in the provenance statement.
+        """
+        build_type = ProvenancePredicate.get_build_type(statement)
+        build_defs: list[ProvenanceBuildDefinition] = [
+            SLSAGithubGenericBuildDefinitionV01(),
+            SLSAGithubActionsBuildDefinitionV1(),
+            SLSANPMCLIBuildDefinitionV2(),
+            SLSAGCBBuildDefinitionV1(),
+            SLSAOCIBuildDefinitionV1(),
+            WitnessGitLabBuildDefinitionV01(),
+        ]
+
+        for build_def in build_defs:
+            if build_def.expected_build_type == build_type:
+                return build_def
+
+        raise ProvenanceError("Unable to find build definition in the provenance statement.")
```

Comment:
A minor improvement could be that we differentiate between 2 cases that we raise an exception:
- There is no build definition, happens when `ProvenancePredicate.get_build_type(statement)` returns None.
- There is a build definition, but we don't have support for its value (the condition `build_def.expected_build_type == build_type` in the for loop never meet).

Right now we only raise one kind of exception message for both cases.

Reply:
See commit [17f3453](https://github.com/oracle/macaron/pull/872/commits/17f3453c4be75784f00a422b71ea829e3f04e4de).